pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,326 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"sigs.k8s.io/jobset/pkg/util/collections"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+)
+
+const (
+	podOwnerKey      = ".metadata.controller"
+	gateDoesNotExist = -1
+)
+
+// PodReconciler reconciles a JobSet object
+type PodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Record record.EventRecorder
+}
+
+func NewPodReconciler(client client.Client, scheme *runtime.Scheme, record record.EventRecorder) *PodReconciler {
+	return &PodReconciler{Client: client, Scheme: scheme, Record: record}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			pod, ok := object.(*corev1.Pod)
+			if !ok || !usingExclusivePlacement(pod) {
+				return false
+			}
+			if !isLeaderPod(pod) {
+				schedulingGateLifted := schedulingGateIndex(pod) == gateDoesNotExist
+				return !schedulingGateLifted
+			}
+			return true
+		})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(r)
+}
+
+func SetupPodReconcilerIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return indexer.IndexField(ctx, &corev1.Pod{}, podOwnerKey, func(obj client.Object) []string {
+		pod := obj.(*corev1.Pod)
+
+		// Make sure the pod is part of a JobSet using exclusive placement.
+		if _, exists := pod.Annotations[jobset.ExclusiveKey]; !exists {
+			return nil
+		}
+
+		// ...make sure it's a Pod owned by a Job...
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil || owner.Kind != "Job" { // owner.APIVersion != batchv1.SchemeGroupVersion.String()
+			return nil
+		}
+		return []string{owner.Name}
+	})
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Get Pod from apiserver.
+	var pod corev1.Pod
+	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+		// we'll ignore not-found errors, since there is nothing we can do here.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If user is not using exclusive placement, we don't need to do anything.
+	topologyKey, exists := pod.Annotations[jobset.ExclusiveKey]
+	if !exists {
+		return ctrl.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("pod", klog.KObj(&pod))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling Pod")
+
```

Comment:
can we add and log the time

```
      start := time.Now()
        defer func() {
                end := time.Since(start).Milliseconds()
                log.Info(fmt.Sprintf("time to reconcile pod %q: %d", pod.Name, end))
        }()
```

Reply:
I think this comment was for debugging the previous design where the logic was all in the reconciler instead of the webhook, so I'm going to close it, but please reopen it if I am mistaken.